python-api/livy/job_handle.py

```diff
@@ -0,0 +1,257 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from concurrent.futures import Future
+import traceback
+import cloudpickle
+from threading import Timer
+import threading
+import base64
+
+# Possible job states.
+PENDING = 'PENDING'
+RUNNING = 'RUNNING'
+CANCELLED = 'CANCELLED'
+FINISHED = 'FINISHED'
+SENT = 'SENT'
+QUEUED = 'QUEUED'
+
+
+class JobHandle(Future):
+
+    """A child class of concurrent.futures.Future. Allows for monitoring and
+        controlling of the running remote job
+
+    """
+    # Poll intervals in seconds
+    _JOB_INITIAL_POLL_INTERVAL = 0.1
+    _JOB_MAX_POLL_INTERVAL = 5
+
+    def __init__(self, conn, session_id, executor):
+        Future.__init__(self)
+        self._conn = conn
+        self._session_id = session_id
+        self._executor = executor
+        self._cancelled = False
+        self._job_id = -1
+        self._done = False
+        self._job_handle_condition = threading.Condition()
+        self._running_callbacks = []
+        self._queued_callbacks = []
+
+    def _start(self, command, serialized_job):
+        self._executor.submit(self._send_job_task, command, serialized_job)
+
+    def _send_job_task(self, command, job):
+        suffix_url = "/" + str(self._session_id) + "/" + command
+        header = {'Content-Type': 'application/json', 'X-Requested-By': 'livy'}
+        job_status = self._conn.send_request('POST', suffix_url,
+                                             headers=header, data=job)
+        self._job_id = job_status.json()['id']
+        self._poll_result()
+
+    def queued(self):
+        """
+        Returns
+        -------
+        True if the job is currently executing.
+        """
+        with self._job_handle_condition:
+            return self._state == QUEUED
+
+    def _invoke_queued_callbacks(self):
+        for callback in self._queued_callbacks:
+            try:
+                callback(self)
+            except Exception:
+                traceback.print_exc()
+            self._queued_callbacks.remove(callback)
+
+    def _invoke_running_callbacks(self):
+        for callback in self._running_callbacks:
+            try:
+                callback(self)
+            except Exception:
+                traceback.print_exc()
+            self._running_callbacks.remove(callback)
+
+    def add_queued_callback(self, fn):
+        """
+        Attaches a callable that will be called when the job state is
+        queued.
+
+        Parameters
+        ----------
+        fn : Function
+            A callable that will be called with this future as its only
+            argument when the job state is queued. The
+            callable will always be called by a thread in the same
```

Comment:
This comment is a little confusing. Why is it needed to specify that the callback will be called in the same process?

Seems to me that specifying which thread will call it is a lot more important.